Add user actions to internal health api indicator responses

jbaiera opened the pull request:
This PR adds some simple diagnosing logic to the ShardsAvailabilityHealthIndicatorService to detect situations where an index's preferred tiers are not compatible with allocation filters on the index settings.

A sample response containing a couple of user actions for the shards availability indicator:

```
GET /_internal/_health/data/shards_availability
```
```json
{
    "cluster_name": "runTask",
    "components": {
        "data": {
            "indicators": {
                "shards_availability": {
                    "status": "red",
                    "summary": "This cluster has 1 unavailable primary.",
                    "details": {
                        "unassigned_replicas": 0,
                        "restarting_primaries": 0,
                        "restarting_replicas": 0,
                        "initializing_primaries": 0,
                        "started_replicas": 0,
                        "initializing_replicas": 0,
                        "unassigned_primaries": 1,
                        "started_primaries": 1,
                        "creating_primaries": 0
                    },
                    "impacts": [
                        {
                            "severity": 1,
                            "description": "Cannot add data to 1 index [test_index_name]. Searches might return incomplete results.",
                            "impact_areas": [
                                "ingest",
                                "search"
                            ]
                        }
                    ],
                    "user_actions": [
                        {
                            "message": "Elasticsearch isn't allowed to allocate some shards from these indices because allocation for those shards has been disabled. Check that the [cluster.routing.allocation.enable] cluster setting is set to [all].",
                            "affected_resources": [
                                "test_index_name"
                            ]
                        },
                        {
                            "message": "Elasticsearch isn't allowed to allocate some shards from these indices because allocation for those shards has been disabled. Check that the [index.routing.allocation.enable] index settings are set to [all].",
                            "affected_resources": [
                                "test_index_name"
                            ]
                        }
                    ]
                }
            }
        }
    }
}
```

jbaiera committed:
Add a basic check for tier preference and allocation filter clashing

jbaiera committed:
Update docs/changelog/85071.yaml

elasticsearchmachine commented:
Hi @jbaiera, I've created a changelog YAML for you.

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -131,16 +148,87 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+        DiscoveryNodeFilters includeFilters = indexMetadata.includeFilters();
+        DiscoveryNodeFilters excludeFilters = indexMetadata.excludeFilters();
+
+        return state.nodes().getDataNodes().values().stream()
+            .filter(node -> {
+                if (node.getRoles().contains(DiscoveryNodeRole.DATA_ROLE)) {
+                    return true;
+                } else {
+                    for (String tierName : tierPreference) {
+                        for (DiscoveryNodeRole role : node.getRoles()) {
+                            if (role.roleName().equals(tierName)) {
+                                return true;
+                            }
+                        }
+                    }
+                }
+                return false;
+            })
+            .filter(node -> {if (requireFilters != null) return requireFilters.match(node); else return true;})
+            .filter(node -> {if (includeFilters != null) return includeFilters.match(node); else return true;})
+            .filter(node -> {if (excludeFilters != null) return excludeFilters.match(node); else return true;})
+            .findAny();
```
This looks like a substantial duplication of logic from the Allocation Explain API - Should we just be using that in this indicator? There are some challenges to obtaining everything we need to call it correctly. Indicators are not injected with components and the explain action is run on the master node. The master node issue is the bigger of the two.

andreidan replied:
I think for some of the deciders (the "simple" ones :) ) it's fine to locally ouput the explain decision and process those decisions (others like the `DiskThresholdDecider` works with settings configured on the master node so we'd need to come up with a different strategy).

For detecting the clash between the "_tier_preference" and "require.data" configuration for routing we don't have to check here if the appropriate tiers are present in the cluster (iterating through nodes and such). We already perform this detection in https://github.com/elastic/elasticsearch/blob/master/x-pack/plugin/ilm/src/main/java/org/elasticsearch/xpack/cluster/metadata/MetadataMigrateToDataTiersRoutingService.java#L563, which is why I was suggesting we could possibly execute a dry_run for this migration service and check if there is an overlap between the indices that need migrating and the indices we detected as unassigned, but a different manual inspection of settings would also work.

I'd argue we should simply output a diagnosis that'll have a resolution of "migrate to data tiers routing".

What do you think?

jbaiera replied:
Looking at the `MetadataMigrateToDataTiersRoutingService` logic for migrating indices, I had some thoughts:

1) The logic presupposes a set of require/include allocation filters. The default is `data` but the user can specify any name for their own that should be migrated. In the case of a custom name other than `data` this detection logic breaks down. We're probably fine in that case, maybe later on we can check filter values and make a guess on a filter to use?
2) The service assumes that the coldest tier is always the first configured. Is that guaranteed?
3) If an index is capable of being migrated, does that necessarily mean that is the reason it is not allocated? I understand that this could be one of two cases: indices are using the "old" default configured filters for tiers or indices are being configured by users with their own hot/cold filters. Are we primarily concerned with just the first case here?

andreidan replied:
Thanks for looking into this Jimmy. You make some great points. 
On:
1. The `data` attribute will likely tackle most cases we see, so I think it's fine to carry on inspecting this one (I think breaching into a blanket solution for _any_ other attribute could overstep over `require._id` which we set when shrinking for e.g., or a user using attributes to create sub-tiers)
2. The service will treat the coldest tier as the correct tier and migrate the settings to the coldest configuration which is where the index should go.
3. We'll likely have to evaluate a chain of deciders output - the ones we evaluate prior to the data tier conflict inspector will rule out _some_ of the other causes. It might not always be the case that this is root cause (it's hard to tell by just looking at a decider) but fixing this will not hurt the cases where there's another underlying issue
4. I don't think I understand what's former and latter in your question? We're concerned by a **mixed** usage of node attributes and tier preference which leads to problems of indices not moving to the desired tier or in certain circumstances unassigned indices.

Hope this helps? Let me know if I missed something

jbaiera replied:
> The service will treat the coldest tier as the correct tier and migrate the settings to the coldest configuration which is where the index should go.

The thing I was wondering about specifically is that the list of tiers is parsed from the preference setting, and the tier name that comes first is assumed to be the coldest, even if it is not actually the coldest. Do we have any guarantees around ensuring that is the case? e.g. if the current logic sees a preference list of `data_hot, data_cold, data_warm` it looks like it would select `data_hot` as the "coldest" tier because it comes first

> I don't think I understand what's former and latter in your question? We're concerned by a mixed usage of node attributes and tier preference which leads to problems of indices not moving to the desired tier or in certain circumstances unassigned indices.

Sorry, I might have been confused about the old way tiering used to work. I couldn't remember if we had a prescribed method of doing tiering that we shipped with all clusters that needed migrating or if it has always been a user configured feature before first class support.

andreidan replied:
>Do we have any guarantees around ensuring that is the case?

We don't validate this at the moment, just that the tiers in the list are valid tier names - maybe open an improvement ticket for this? 
I haven't seen any configuration like this so far so I think for now it's fine to move ahead with the assumption the first is the coldest.

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -131,16 +148,87 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+    private static Diagnosis diagnoseReason(ShardRouting routing, ClusterState state) {
+        return Optional.ofNullable(state.metadata().index(routing.index()))
+            .map(indexMetadata -> {
+                // Check: Data Tier conflicts with required attributes
```
Currently we're only checking for data tier preference agreeing with allocation filtering - This is trying to detect a well known allocation snag that users experience but it's not necessarily the _cause_ of a shard being unallocated, just something we know will keep it from being assigned.

jbaiera reviewed (commented):
Some initial thoughts that I'm seeking feedback on

jbaiera committed:
Add action data model to indicator results

jbaiera committed:
add user actions to indicator result output

jbaiera committed:
Switch to collection on UserAction for affected resources

jbaiera committed:
WIP generating a few actions for shards availability indicator

jbaiera committed:
Change UserAction affectedResources to just be a set

jbaiera committed:
Allow exporting of the allocation deciders from the cluster module.

Normally the way you would get these deciders is to have them injected, but the health indicators are created at the start of the node and do not have their dependencies injected.

jbaiera committed:
Add allocation service and deciders to the health indicator

jbaiera committed:
Rewrite the diagnosis logic with allocation decisions.

Indices can be placed in multiple actions, so don't short circuit.
This way we do not have to reference the allocators directly by instance, but can still do so generically by name.

jbaiera committed:
Merge branch 'master' into health-api-shard-unassigned-actions

jbaiera committed:
Fix compilation after merge

jbaiera committed:
Add additional shard diagnosis checks

jbaiera committed:
Remove searchable snapshot user actions for now

jbaiera committed:
Remove id from user action serialization for now

jbaiera committed:
Get tests happy once again

jbaiera committed:
Remove searchable snapshot related user actions

jbaiera committed:
Add first tests for indicator user actions

jbaiera committed:
Update the diagnosis logic to avoid false results on missing decisions

jbaiera committed:
Finish unit tests for diagnosis checks

jbaiera committed:
Simplify user action creation and tracking

jbaiera committed:
Merge branch 'master' into health-api-shard-unassigned-actions

jbaiera committed:
Add user actions to serialization tests

jbaiera committed:
Add support for skipping processing unless drilling down

jbaiera committed:
Add rest test for user actions

jbaiera committed:
Add user action messages and clean up.

jbaiera committed:
null out missing help urls

jbaiera committed:
Satisfy precommit

jbaiera committed:
Merge remote-tracking branch 'jbaiera/health-api-shard-unassigned-actions' into health-api-shard-unassigned-actions

jbaiera commented on server/src/main/java/org/elasticsearch/health/HealthIndicatorResult.java:
```diff
@@ -20,9 +21,21 @@ public record HealthIndicatorResult(
     HealthIndicatorDetails details,
-    List<HealthIndicatorImpact> impacts
+    List<HealthIndicatorImpact> impacts,
+    @Nullable List<UserAction> userActions
```
I'm on the fence between making this nullable (so it doesn't show up in the output if there are none) vs making it always there but empty (for ease of use in the code)

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+    public static final UserAction.Definition ACTION_RESTORE_FROM_SNAPSHOT = new UserAction.Definition(
+        "restore_from_snapshot",
+        Explanations.Allocation.NO_COPIES,
+        null
+    );
+    public static final UserAction.Definition ACTION_ENABLE_ALLOCATIONS = new UserAction.Definition(
+        "enable_allocations",
+        "Elasticsearch isn't allowed to allocate shards from these indices because allocation for those shards has been disabled. "
+            + "Check that the ["
+            + EnableAllocationDecider.INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] index settings and the ["
+            + EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] cluster setting are not set to ["
+            + EnableAllocationDecider.Allocation.NONE.toString().toLowerCase(Locale.getDefault())
+            + "].",
+        null
+    );
+    public static final UserAction.Definition ACTION_SHARD_LIMIT = new UserAction.Definition(
+        "increase_shard_limit",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because each node in the tier has "
+            + "reached its shard limit. Increase the values for the ["
+            + ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.getKey()
+            + "] index setting on each index or add more nodes to the cluster.",
+        null
+    );
+    public static final UserAction.Definition ACTION_MIGRATE_TIERS = new UserAction.Definition(
+        "migrate_data_tiers",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because no nodes in the tier are "
+            + "compatible with the allocation filters in the index settings. Remove the conflicting allocation filters from each index's "
+            + "settings or try migrating to data tiers using the data tier migration action.",
+        null
+    );
+    public static final UserAction.Definition ACTION_INCREASE_TIER_CAPACITY = new UserAction.Definition(
+        "increase_tier_capacity_for_allocations",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because there are not enough nodes "
+            + "in the tier to allocate each shard copy on a different node. Increase the number of nodes in this tier or decrease the "
+            + "number of replicas your indices are using.",
+        null
+
+    );
```
Looking for specific feedback on the messages displayed in the user actions here. Help URL's are currently null while we finalize where they'll be.

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+    // TODO: #85572 Fill in help URLs once they are finalized
+    public static final UserAction.Definition ACTION_RESTORE_FROM_SNAPSHOT = new UserAction.Definition(
+        "restore_from_snapshot",
+        Explanations.Allocation.NO_COPIES,
```
For ease of review on Github this explanation text is as follows:

> Elasticsearch can't allocate this shard because there are no copies of its data in the cluster. Elasticsearch will allocate this shard when a node holding a good copy of its data joins the cluster. If no such node is available, restore this index from a recent snapshot.

jbaiera reviewed (commented):
Some specific feedback I'd like to gather:

elasticmachine commented:
Pinging @elastic/es-data-management (Team:Data Management)

Leaf-Lin commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+            + "settings or try migrating to data tiers using the data tier migration action.",
+        null
+    );
+    public static final UserAction.Definition ACTION_INCREASE_TIER_CAPACITY = new UserAction.Definition(
+        "increase_tier_capacity_for_allocations",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because there are not enough nodes "
+            + "in the tier to allocate each shard copy on a different node. Increase the number of nodes in this tier or decrease the "
+            + "number of replicas your indices are using.",
```
Should we also suggest enabling autoscaling and deleting unwanted data?

jbaiera replied:
This user action is currently surfaced when we have a shard that can't be placed anywhere because it would be hosted on a node that already holds the shard. I don't know if deleting unwanted data would solve this, unless you mean removing the entire index?

I took a look and it seems that autoscaling ignores this decider because it mostly deals with allocating replicas. I'll try refining the text based on that.

Leaf-Lin commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+        "migrate_data_tiers",
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because no nodes in the tier are "
+            + "compatible with the allocation filters in the index settings. Remove the conflicting allocation filters from each index's "
+            + "settings or try migrating to data tiers using the data tier migration action.",
```
Should we add the actual API to run in the action message? `POST  /_ilm/migrate_to_data_tiers`

jbaiera replied:
Added in 3370adf

andreidan commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -82,27 +110,78 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+        "Elasticsearch isn't allowed to allocate this shard to any of the nodes in its data tier because each node in the tier has "
+            + "reached its shard limit. Increase the values for the ["
+            + ShardsLimitAllocationDecider.INDEX_TOTAL_SHARDS_PER_NODE_SETTING.getKey()
+            + "] index setting on each index or add more nodes to the cluster.",
```
```suggestion
            + "] index setting on each index or add more nodes to the target tier.",
```

jbaiera replied:
fixed with 3a7c310

jbaiera committed:
"add nodes to cluster" -> "add nodes to target tiers"

jbaiera committed:
Refine message text, add migrate data tiers action to message.

jbaiera committed:
Add a user action that says to enable data tiers if they are required but none are found.

jbaiera committed:
Add default action to check allocation explain if cannot diagnose decider problem.

jbaiera committed:
Precommit formatting

jbaiera committed:
Merge branch 'master' into health-api-shard-unassigned-actions

jbaiera committed:
Fix compile issues after update

andreidan commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+        return false;
+    }
+
+    private boolean checkDataTierRelatedIssues(
```
Shall we document? 

Also for these private methods, should they be package default visible and unit tested? I like the tests you added already, however going through the indicator covers many of these methods in one unit, so if a unit test fails it's not immediately clear where to look to debug.
What do you think?

jbaiera replied:
I reworked the unit tests in [d75e324](https://github.com/elastic/elasticsearch/pull/85071/commits/d75e324e7af06e7931fc00316bd2c04616eb496d). I agree, they needed some simplification and focusing on just the diagnosis calls.

andreidan commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+    ) {
+        boolean diagnosisFound = false;
+        if (indexMetadata.getTierPreference().size() > 0) {
+            List<NodeAllocationResult> dataTierNodes = nodeAllocationResults.stream()
```
This was a bit confusing to me as for `dataTierNodes` I would've expected a `List<DiscoveryNode>`. Would this be better named `dataTierAllocationResults` or something like that?

andreidan commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+                }
+
+                // All tier nodes conflict with allocation filters?
+                if (dataTierNodes.stream().allMatch(nodeHasDeciderResult(FilterAllocationDecider.NAME, Decision.Type.NO))) {
```
I believe this is too broad? Would this catch the `require._id` filtering we use when shrinking for example?
Also I think we can have the reverse case too, where the `require.data` filtering says YES and but the target tier says `NO` (?) 

I have a feeling we'd be better with simulating a migration and detecting if any index in the cluster needs to be migrated ie. `MigrateToDataTiersRouting#migrateToDataTiersRouting( ..., dryRun true)`. What do you think? 

@dakrone

jbaiera replied:
I agree, giving it a dry run through the migrate action would be good, but it does live in the ILM project. It would either need to be moved to server, have its logic duplicated, or a simpler check performed.

dakrone replied:
> I believe this is too broad? Would this catch the `require._id` filtering we use when shrinking for example?

I think this is too broad also, I wonder if instead of making this a blanket "filter allocation said NO" thing we could treat it similar to the migrate API where we special-case the "data" attribute? I don't think we want to say "run the migrate API" when an index happens to be yellow due to `_id` pinning required for the shrink API.

dakrone replied:
I don't know what granularity we have in the code, but we might want to add another check for "oops you fat-fingered a filtered attribute and there is no node in the cluster that has that attribute" if possible. (not necessarily in this PR, maybe in future work?)

jbaiera replied:
Added [c671eb5](https://github.com/elastic/elasticsearch/pull/85071/commits/c671eb5b0da7c3ae963637e8b509a68c4ab30875) to not only check for old data attribute filters but also to see if they keep nodes from being selected (since their existence does not automatically explain an unassigned shard)

jbaiera replied:
++ to the "no nodes match filter" check, but definitely should be after this lands.

andreidan commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -56,12 +72,24 @@
+    private static final String DATA_TIER_ALLOCATION_DECIDER_NAME = "data_tier";
+
     private final ClusterService clusterService;
+    private final AllocationService allocationService;
+    private final AllocationDeciders allocationDeciders;
```
Would it make sense to have this diagnosis extracted into a service of its own?

jbaiera replied:
This is an interesting idea. I'm inclined to disagree for now because we only make use of this logic in the health indicator whereas I think a separate service should be more reusable. That said, the indicator code is getting filled up with diagnosis methods and getting harder to read. Maybe it makes sense to do this as a follow up so it's easier to throw out if we don't like how it looks?

andreidan reviewed (changes requested):
Thanks for working on this Jimmy. 

I left a few suggestions and questions.

jbaiera committed:
Simplify the diagnose method call signatures

jbaiera committed:
Avoid creating a list on every method call.

jbaiera committed:
Add javadoc for internal diagnosis methods.

jbaiera committed:
Fully integrate user actions into indicator results

jbaiera committed:
rename dataTierNodes -> dataTierAllocationResults

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+            + EnableAllocationDecider.INDEX_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] index settings and the ["
+            + EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] cluster setting are not set to ["
+            + EnableAllocationDecider.Allocation.NONE.toString().toLowerCase(Locale.getDefault())
```
Can we change this to be something like "Check that the [...] index settings and [...] cluster settings are set to [all]"?

Also, +1 to what Andrei said below about trying to determine whether it's a cluster-level or index-level in the message.

jbaiera replied:
Applied in [bbd994d](https://github.com/elastic/elasticsearch/pull/85071/commits/bbd994dd652e9ea496ecca48a894318852a93996)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+    );
+    public static final UserAction.Definition ACTION_ENABLE_TIERS = new UserAction.Definition(
+        "enable_data_tiers",
+        "Elasticsearch isn't allowed to allocate shards from these indices because the indices expect to be allocated to data tier nodes, "
+            + "but none were found in the cluster. Add nodes with data tier roles ("
+            + DataTier.DATA_HOT
+            + ", "
+            + DataTier.DATA_WARM
+            + ", "
+            + DataTier.DATA_COLD
+            + ", etc.) to the cluster.",
```
Yeah, we should find the exact tiers required, because it could be they're missing `DataTier.DATA_CONTENT` (or some other tier we don't mention here)

jbaiera replied:
Applied in [db033ef](https://github.com/elastic/elasticsearch/pull/85071/commits/db033efcf8f0c16905bdba2162edaa785a645632)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+    public static final UserAction.Definition ACTION_SHARD_LIMIT = new UserAction.Definition(
+        "increase_shard_limit",
+        "Elasticsearch isn't allowed to allocate some shards from these indices to any of the nodes in its data tier because each node in "
+            + "the tier has reached its shard limit. Increase the values for the ["
```
Can we include what the shard limit is in the message?

jbaiera replied:
This is a bit of a weird one because of how the messages are currently laid out. The message displayed to a user is part of an immutable action definition. If we include an index scope setting value in the message for the user action we would need to create a new user action entry for each index affected by this issue. This would be the only way to have each message carry the value for each index.

Furthermore, if the number of indices that this problem affects is large, the response would be substantially larger if each affected index gets its own user action block.

andreidan replied:
I think it'll usually be the case of one tier needing a scale-up. Could we specify which tier(s) need it? The current message is too generic to be actionable IMO

andreidan replied:
Also, could you please list a complete response in the PR description? (that'd be a life saver when processing all the diagnosis we do here :) - and we could lift and shift that into the docs)

jbaiera replied:
Tier scaling change in [6042c00](https://github.com/elastic/elasticsearch/pull/85071/commits/6042c00148846e3c40fd2976567ee283489460b8) and an example of a response was added to the PR description - I'll work on getting the remaining user actions added in as well

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
         private int started = 0;
         private int relocating = 0;
-        private Set<String> indicesWithUnavailableShards = new HashSet<>();
+        private final Set<String> indicesWithUnavailableShards = new HashSet<>();
```
This is random and I'm missing context, but would it be helpful to have this be a `TreeSet` so that the indices are always in a consistent order?

jbaiera replied:
It looks like this field is only pulled up in `getImpacts` and that method sorts the indices by priority, then name. Probably don't need to have them presorted?

That said, It might be good to have the indices sorted the same way in the user action output.

jbaiera replied:
Added in [422d39b](https://github.com/elastic/elasticsearch/pull/85071/commits/422d39b22890077784efaf2c90c6a64f95985bdc)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+            .anyMatch(decision -> deciderName.equals(decision.label()) && outcome == decision.type());
+    }
+
+    private boolean checkIsAllocationDisabled(
```
I think these methods should return the definition/routing rather than take the consumer and do a builder-y type of buildup. They're much easier to think about because you can see exactly where the output is used rather than passing a consumer through that you have to go look up its actual function.

jbaiera replied:
very much agreed - refactoring is in [948b5e2](https://github.com/elastic/elasticsearch/pull/85071/commits/948b5e29b176071ffedb7613fa99563bb2d213e9), [ec52fca](https://github.com/elastic/elasticsearch/pull/85071/commits/ec52fca4b95cb98e7016272d465b42bf48cf489d)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -264,6 +499,27 @@ public List<HealthIndicatorImpact> getImpacts() {
             return impacts;
         }
 
+        public List<UserAction> getUserActions(boolean includeDetails) {
```
Can you add javadoc for this method?

jbaiera replied:
Added in [1cdba6f](https://github.com/elastic/elasticsearch/pull/85071/commits/1cdba6f11ed80f91d470a1631a73343262cd9bf5)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/ClusterModule.java:
```diff
@@ -337,6 +337,10 @@ private static ShardsAllocator createShardsAllocator(
         return Objects.requireNonNull(allocatorSupplier.get(), "ShardsAllocator factory for [" + allocatorName + "] returned null");
     }
 
+    public AllocationDeciders getAllocationDeciders() {
+        return allocationDeciders;
+    }
```
Rather than add this at the `ClusterModule` level, we could add it to the actual `AllocationService` (since it has access to the deciders internally in the first place). That might make more sense since the deciders are more part of the allocation service than the cluster module; what do you think?

jbaiera replied:
Fixed in [3819d46](https://github.com/elastic/elasticsearch/pull/85071/commits/3819d460d24a726a25bb8e27880522065b95c21a)

jbaiera committed:
Simplify unit tests to only target diagnosis methods.

jbaiera committed:
Sort indices in user action output

jbaiera committed:
Add javadoc to getUserActions

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
 
-    private static class ShardAllocationStatus {
+    private void diagnoseUnassigned(
+        BiConsumer<UserAction.Definition, ShardRouting> diagnosisOutput,
```
I changed these methods to take a list as an argument and add the definition to the list if it is applicable instead.

Changes in [948b5e2](https://github.com/elastic/elasticsearch/pull/85071/commits/948b5e29b176071ffedb7613fa99563bb2d213e9), [ec52fca](https://github.com/elastic/elasticsearch/pull/85071/commits/ec52fca4b95cb98e7016272d465b42bf48cf489d)

jbaiera committed:
Get AllocationDeciders from the AllocationService instead of ClusterModule.

jbaiera committed:
Cleanup warning

jbaiera committed:
Split enable allocation action into two based on which setting should be changed.

jbaiera committed:
Add exhaustive list of actions for adding data tiers

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +255,133 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+                .collect(Collectors.toList());
+            if (dataTierNodes.isEmpty()) {
+                diagnosisOutput.accept(ACTION_ENABLE_TIERS, shardRouting);
+                diagnosisFound = true;
```
If none of the DiscoveryNodes are marked with any tier roles, I don't think that method will return any tiers. This case is likely to be rare since a combination of the `data` role is acceptable to allocate an index from any tier, not to mention tier preference accepts a list of tier roles specifically to avoid this problem if an entire tier is missing.

I think with all that in mind it would be safe to simply iterate over the preferred tiers list on the index and mark each tier found as needing to be enabled on the cluster.

jbaiera replied:
Applied in [db033ef](https://github.com/elastic/elasticsearch/pull/85071/commits/db033efcf8f0c16905bdba2162edaa785a645632)

jbaiera committed:
Precommit precommit precommit

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -83,27 +111,99 @@ public HealthIndicatorResult calculate(boolean includeDetails) {
+            + EnableAllocationDecider.CLUSTER_ROUTING_ALLOCATION_ENABLE_SETTING.getKey()
+            + "] cluster setting are not set to ["
+            + EnableAllocationDecider.Allocation.NONE.toString().toLowerCase(Locale.getDefault())
+            + "].",
```
Applied in [bbd994d](https://github.com/elastic/elasticsearch/pull/85071/commits/bbd994dd652e9ea496ecca48a894318852a93996)

jbaiera committed:
Add checks when filter allocation fails to see if data tier migration should be done.

jbaiera committed:
Fix test

jbaiera committed:
precommit grumble grumble

jbaiera commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into health-api-shard-unassigned-actions

jbaiera committed:
Split shard limit diagnosis out by data tier

jbaiera committed:
Clear out some warnings in the test class

jbaiera committed:
Fix failing rest test

jbaiera committed:
Make the migrate to data tiers action more specific about the problem

jbaiera committed:
Add more information to increase shard limit actions

jbaiera committed:
Add more information to same shard allocation decider rejections

andreidan reviewed (commented):
Thanks for iterating on this @jbaiera 

I believe we're nearly there 🚀 Left a few rather minor suggestions/questions

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +358,292 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+     * @param state Current cluster state.
+     * @param nodeAllocationResults A list of results for each node in the cluster from the allocation explain api
+     */
+    void diagnoseAllocationResults(
```
I originally set it up to return lists but it was a lot of creating lists just to return them and have their contents added to the list of the calling method. I switched to this consuming model to cut down on list allocations, but maybe it's not worth the tradeoff for readability?

andreidan replied:
Would it be 2 collections? for the `checkIsAllocationDisabled` and `checkDataTierRelatedIssues` methods (as `diagnoseUnassignedShardRouting` just delegates). If readability is improved and testability is easier I'd say it's worth it, but I'll leave it up to you.

jbaiera replied:
Fixed in [fc07c71](https://github.com/elastic/elasticsearch/pull/85071/commits/fc07c71ca898cb8c49da6e3916708694988d7fef)

jbaiera committed:
Return lists from diagnose methods

jbaiera committed:
Simplify some collections to be reused

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +358,292 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+     * @param shardRouting The shard routing that is unassigned with a last status of DECIDERS_NO
+     * @param state Current cluster state
+     */
+    private void explainAllocationsAndDiagnoseDeciders(List<UserAction.Definition> actions, ShardRouting shardRouting, ClusterState state) {
```
Fixed in [fc07c71](https://github.com/elastic/elasticsearch/pull/85071/commits/fc07c71ca898cb8c49da6e3916708694988d7fef)

jbaiera commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +358,292 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+                    Optional<String> preferredTier = indexMetadata.getTierPreference()
+                        .stream()
+                        .filter(dataTierRolesAvailable::contains)
+                        .findFirst();
```
Fixed in [b0263a1](https://github.com/elastic/elasticsearch/pull/85071/commits/b0263a1c97df0887940891e45fa3cffdede8c4d6)

andreidan reviewed (approved):
Thanks for iterating on this Jimmy (the tests are 👌 ❤️)

LGTM 🚀 

@dakrone any amendments on your end?

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -146,7 +358,281 @@ private static boolean isUnassignedDueToNewInitialization(ShardRouting routing)
+     * @param clusterState the current cluster state.
+     * @return a list of user actions to take.
+     */
+    List<UserAction.Definition> checkDataTierRelatedIssues(
+        IndexMetadata indexMetadata,
+        List<NodeAllocationResult> nodeAllocationResults,
+        ClusterState clusterState
+    ) {
```
Is there a way we can break this method up? It's very difficult to read as it's well over 100 lines long. Maybe we can put the `ShardsLimitAllocationDecider` branch into a method (as well as the other self-contained `if` statements under the top-level `else` clause?

jbaiera replied:
fixed in [f5bf091](https://github.com/elastic/elasticsearch/pull/85071/commits/f5bf0915a73f8f893111249d7e5985ae062fd1c5)

dakrone commented on server/src/main/java/org/elasticsearch/cluster/routing/allocation/ShardsAvailabilityHealthIndicatorService.java:
```diff
@@ -264,20 +750,64 @@ public List<HealthIndicatorImpact> getImpacts() {
+            .limit(maxIndices)
+            .collect(joining(", "));
         if (maxIndices < indices.size()) {
             truncatedIndicesString = truncatedIndicesString + ", ...";
```
I am halfway tempted to suggest we replace this with "…" but I think that would be a bit too far :D

dakrone reviewed (approved):
LGTM also, I left only really minor stuff.

jbaiera committed:
Split out data tier diagnosis methods further.

Do not use Optional for method parameters.

jbaiera committed:
Fix subtle testing bug
